### PR TITLE
feat: allow configurable retryStrategy

### LINKED ISSUE #1951: Expose Retry Count and Retry Timeout For Sink
# Summary

Currently, we have infinite retry with a fixed interval of 1 ms, which can lead to problems when the sink is in a bad state. We should provide an option for the user to configure max retry timeout interval. 

We should also expose Retry Count so the user can either drop the message based on the number of retries or write to the fallback sink.

Note: Retry is an optimistic retry. The max retry can be >= the user-configured retry count because the counter will be reset each time the pod is restarted.

# Use Cases

When would you use this?
* Users want to drop the message after 'N' attempts.
* Users don't want to write retry logic.

---

```[tasklist]
### Tasks
- [ ] Add retryStrategy to Sink CRD
- [ ] Fixed retry for Pipeline
- [ ] Fixed retry for MonoVertex
- [ ] Exponential retry for Pipeline
- [x] ~Exponential retry for MonoVertex~
- [ ] https://github.com/numaproj/numaflow/issues/2016
- [ ] Expose retry count from sinks
```

### PR DESCRIPTION
**Pull request by kohlisid**
fixes #1951 

For testing refer - https://github.com/numaproj/numaflow/issues/2016

### CODE REVIEW

File: pkg/apis/numaflow/v1alpha1/retry_strategy.go [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+type OnFailRetryStrategy string
```
Comment (whynowy) [inline pkg/apis/numaflow/v1alpha1/retry_strategy.go:20]: OnFailureRetryStrategy

File: pkg/apis/numaflow/v1alpha1/retry_strategy.go [hunk 2]
```diff
@@ -0,0 +1,58 @@
+/*
```
Comment (whynowy) [inline pkg/apis/numaflow/v1alpha1/retry_strategy.go:1]: This header is different from others.

File: pkg/sinks/forward/forward.go [hunk 3]
```diff
@@ -357,76 +357,120 @@ func (df *DataForward) ackFromBuffer(ctx context.Context, offsets []isb.Offset)
...
-				if errors.Is(err, &udsink.WriteToFallbackErr) && df.opts.fbSinkWriter != nil && !isFbSinkWriter {
-					fallbackMessages = append(fallbackMessages, msg)
-					continue
-				}
+		err = wait.ExponentialBackoffWithContext(ctx, backoffCond, func(_ context.Context) (done bool, err error) {
+			// Note: this is an unwanted memory allocation during a happy path. We want only minimal allocation
+			// since using failedMessages is an unlikely path.
+			var failedMessages []isb.Message
+			needRetry := false
+			df.opts.logger.Info("MYDEBUG: Trying sink write at: ", time.Now().String())
```
Comment (vigith) [inline pkg/sinks/forward/forward.go:385]: remove these?

File: pkg/sinks/forward/forward.go [hunk 4]
```diff
@@ -357,76 +357,120 @@ func (df *DataForward) ackFromBuffer(ctx context.Context, offsets []isb.Offset)
...
+		// Check if we still have messages left to be processed
+		if len(messagesToTry) > 0 {
+			df.opts.logger.Infof("%d tries exhausted in sink, messagesLeft %d, Next strategy %s",
+				backoffCond.Steps, len(messagesToTry), failStrategy)
+			// Check what is the failure strategy to be followed after retry exhaustion
+			switch failStrategy {
+			case dfv1.OnFailRetry:
+				// If on failure, we keep on retrying then lets continue the loop and try all again
+				continue
+			case dfv1.OnFailFallback:
```
Comment (vigith) [inline pkg/sinks/forward/forward.go:455]: add metrics for Fallback?

File: rust/monovertex/src/config.rs [hunk 5]
```diff
@@ -19,8 +19,9 @@ const DEFAULT_LAG_CHECK_INTERVAL_IN_SECS: u16 = 5;
 const DEFAULT_LAG_REFRESH_INTERVAL_IN_SECS: u16 = 3;
 const DEFAULT_BATCH_SIZE: u64 = 500;
 const DEFAULT_TIMEOUT_IN_MS: u32 = 1000;
-const DEFAULT_MAX_SINK_RETRY_ATTEMPTS: u16 = 10;
-const DEFAULT_SINK_RETRY_INTERVAL_IN_MS: u32 = 1;
+pub const DEFAULT_MAX_SINK_RETRY_ATTEMPTS: u16 = u16::MAX - 1;
+pub const DEFAULT_SINK_RETRY_INTERVAL_IN_MS: u32 = 1;
```
Comment (vigith) [inline rust/monovertex/src/config.rs:23]: can we avoid this to be `pub`?

File: rust/monovertex/src/config.rs [hunk 6]
```diff
@@ -117,9 +120,33 @@ impl Settings {
             settings.is_fallback_enabled = mono_vertex_obj
                 .spec
                 .sink
+                .clone()
                 .ok_or(Error::ConfigError("Sink not found".to_string()))?
                 .fallback
                 .is_some();
+
+            if let Some(retry_strategy) = mono_vertex_obj.spec.sink.clone().unwrap().retry_strategy
```
Comment (vigith) [inline rust/monovertex/src/config.rs:128]:
```suggestion
            if let Some(retry_strategy) = mono_vertex_obj.spec.sink.clone().expect("sink should not be empty").retry_strategy
```

**Review by vigith — COMMENTED** [review]
need to review rust code, just skimmed over. will review after tests are added.

File: pkg/sinks/forward/forward.go (hunk 4 above)
Comment (kohlisid) [reply pkg/sinks/forward/forward.go]: Added seperate metrics for FB on vertex and monovertex

File: pkg/metrics/metrics.go [hunk 7]
```diff
@@ -86,7 +86,7 @@ var (
 		Help:      "Total number of bytes written",
 	}, []string{LabelVertex, LabelPipeline, LabelVertexType, LabelVertexReplicaIndex, LabelPartitionName})
 
-	// WriteMessagesError is used to indicate the number of errors messages written
+	// WriteMessagesError is used to indicate the number of errors while writing to a sink
```
Comment (whynowy) [inline pkg/metrics/metrics.go:89]: Why is `writing to a sink`?

File: pkg/sinks/sink.go [hunk 8]
```diff
@@ -246,6 +246,14 @@ func (u *SinkProcessor) Start(ctx context.Context) error {
 			forwardOpts = append(forwardOpts, sinkforward.WithCallbackUploader(cbPublisher))
 		}
 
+		// Derive the retryStrategy from the spec and add to the forwarder if it is valid
+		retryStrategy := u.VertexInstance.Vertex.Spec.Sink.GetRetryStrategy()
+
+		if ok := u.VertexInstance.Vertex.Spec.Sink.IsValidSinkRetryStrategy(retryStrategy); ok != nil {
```
Comment (whynowy) [inline pkg/sinks/sink.go:252]: These functions do not make sense - you already have the sink object, you get the retry strategy from the object, and then call a function of the sink object to tell if the strategy is valid...

Comment (kohlisid) [reply pkg/sinks/sink.go:252]: My bad, oversight.
Have updated to check for validation while fetching retryStrategy

File: pkg/apis/numaflow/v1alpha1/vertex_types.go [hunk 9]
```diff
@@ -643,7 +643,7 @@ func (av AbstractVertex) HasUDTransformer() bool {
 }
 
 func (av AbstractVertex) HasFallbackUDSink() bool {
-	return av.IsASink() && av.Sink.Fallback != nil && av.Sink.Fallback.UDSink != nil
+	return av.IsASink() && av.Sink.hasValidFallbackSink()
```
[the remark this answers is not in the document]
Comment (kohlisid) [reply pkg/apis/numaflow/v1alpha1/vertex_types.go:646]: `hasValidFallbackSink()` function just checks those original two conditions together, the same logic. 

```
// hasValidFallbackSink checks if the Sink vertex has a valid fallback sink configured
func (s *Sink) hasValidFallbackSink() bool {
	return s.Fallback != nil && s.Fallback.UDSink != nil
}
```

File: pkg/sinks/forward/forward.go [hunk 10]
```diff
@@ -357,82 +357,190 @@ func (df *DataForward) ackFromBuffer(ctx context.Context, offsets []isb.Offset)
...
+// updateSinkWriteMetrics updates metrics related to data writes to a sink.
+// Metrics are updated based on whether the operation involves the primary or fallback sink.
+func (df *DataForward) updateSinkWriteMetrics(writeCount int, writeBytes float64, sinkWriterName string, isFallback bool) {
+	// Use default metrics for primary sink operations
+	writeCountMetric := metrics.WriteMessagesCount
+	writeBytesMetric := metrics.WriteBytesCount
+
+	// Assign fallback-specific metrics for both write count and write bytes
+	// if this is for Fallback Sink
+	if isFallback {
```
[the remark this answers is not in the document]
Comment (kohlisid) [reply pkg/sinks/forward/forward.go:504]: So you want WriteMessages to be the total for fallback + primary?
I was keeping this for primary sink and fallback metric for the other sink separately

File: pkg/apis/numaflow/v1alpha1/sink.go [hunk 11]
```diff
@@ -134,3 +139,20 @@ func (s Sink) getFallbackUDSinkContainer(mainContainerReq getContainerReq) corev
...
 }
+
+// hasValidFallbackSink checks if the Sink vertex has a valid fallback sink configured
+func (s *Sink) hasValidFallbackSink() bool {
+	return s.Fallback != nil && s.Fallback.UDSink != nil
+}
+
+// HasValidSinkRetryStrategy checks if the provided RetryStrategy is valid based on the sink's configuration.
+// This validation ensures that the retry strategy is compatible with the sink's current setup
+func (s Sink) HasValidSinkRetryStrategy() error {
```
[the remark this answers is not in the document]
Comment (kohlisid) [reply pkg/apis/numaflow/v1alpha1/sink.go:150]: Can move to validation

File: pkg/sinks/sink.go [hunk 12]
```diff
@@ -246,6 +246,9 @@ func (u *SinkProcessor) Start(ctx context.Context) error {
 			forwardOpts = append(forwardOpts, sinkforward.WithCallbackUploader(cbPublisher))
 		}
 
+		// Derive the retryStrategy from the spec and add to the forwarder
+		forwardOpts = append(forwardOpts, sinkforward.WithRetryStrategy(u.VertexInstance.Vertex.Spec.Sink.RetryStrategy))
```
Comment (whynowy) [inline pkg/sinks/sink.go:250]: Minor: we pass `u.VertexInstance` to `NewDataForward()`, which means everything in the spec is available there.

File: pkg/sinks/forward/forward.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (whynowy) [reply pkg/sinks/forward/forward.go:504]: yes. or we have to sum these 2 metrics to know the total writes for sinks.

File: pkg/sinks/forward/forward.go [hunk 13]
```diff
@@ -443,3 +551,20 @@ func errorArrayToMap(errs []error) map[string]int64 {
...
+		return wait.Backoff{
+			Duration: dfv1.DefaultRetryInterval,
+			// Always account for the first try
+			Steps: dfv1.DefaultRetrySteps + 1,
+		}, dfv1.OnFailureRetry
+	}
+	// Initial interval duration and number of retries are taken from DataForward settings.
+	backoff := df.opts.retryStrategy.GetBackoff()
+	// Adding a step for the first try
+	backoff.Steps = backoff.Steps + 1
```
Comment (whynowy) [inline pkg/sinks/forward/forward.go:568]: If we need to do this trick, why don't we correct the definition in the spec?

Comment (kohlisid) [reply pkg/sinks/forward/forward.go:568]: I had thought about it, just one corner case where the user gives 0 somehow in the spec. Then there will be no sink writes happening.
This just ensures that we try `atleast` once, regardless of what the user gives us. 
I couldn't think a valid case where the user would not want that.

File: rust/monovertex/src/forwarder.rs [hunk 14]
```diff
@@ -222,85 +224,140 @@ impl Forwarder {
...
+                    }
+                }
+            }
+            break;
         }
 
+        let fallback_msgs_count = fallback_msgs.len() as u64;
         // If there are fallback messages, write them to the fallback sink
-        if !fallback_msgs.is_empty() {
+        if fallback_msgs_count > 0 {
```
Comment (vigith) [inline rust/monovertex/src/forwarder.rs:286]: why did you change this?

Comment (kohlisid) [reply rust/monovertex/src/forwarder.rs:286]: `let fallback_msgs_count = fallback_msgs.len() as u64;`
I had extracted the count because I had to use it for metrics, have updated there. So can revert this now

File: rust/monovertex/src/forwarder.rs [hunk 15]
```diff
@@ -222,67 +223,43 @@ impl Forwarder {
...
-                    warn!(
-                        "Retry attempt {} due to retryable error. Errors: {:?}",
-                        attempts, error_map
-                    );
-                    sleep(tokio::time::Duration::from_millis(
-                        config().sink_retry_interval_in_ms as u64,
-                    ))
+        // only breaks out of this loop based on the retry strategy unless all the messages have been written to sink
+        // successfully.
+        loop {
```
Comment (vigith) [inline rust/monovertex/src/forwarder.rs:228]: remove this loop as there are mutually exclusion operations happening inside the same loop.

File: rust/monovertex/src/config.rs [hunk 16]
```diff
@@ -17,8 +17,9 @@ const DEFAULT_LAG_CHECK_INTERVAL_IN_SECS: u16 = 5;
 const DEFAULT_LAG_REFRESH_INTERVAL_IN_SECS: u16 = 3;
 const DEFAULT_BATCH_SIZE: u64 = 500;
 const DEFAULT_TIMEOUT_IN_MS: u32 = 1000;
-const DEFAULT_MAX_SINK_RETRY_ATTEMPTS: u16 = 10;
+const DEFAULT_MAX_SINK_RETRY_ATTEMPTS: u16 = u16::MAX;
 const DEFAULT_SINK_RETRY_INTERVAL_IN_MS: u32 = 1;
+const DEFAULT_SINK_RETRY_ON_FAIL_STRATEGY: &str = "retry";
```
Comment (vigith) [inline rust/monovertex/src/config.rs:22]: convert this to a ENUM

**Review by vigith — APPROVED** [review]
feel free to create follow up PRs for these.

**Review by whynowy — APPROVED** [review]
